File: server/script_packages/mypackage/fun/ms.py
import os
from mypackage.module import WebDriverWait, EC, By
from random import seed, randint
import re
import time
import json


def get_sku_items_num(driver, sku_item_link):

    result = {'total_num': 0, 'num_per_page': 0}
    driver.get(sku_item_link)

    try:
        totalProductsOutput = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located(
                (By.CSS_SELECTOR, "p[class='c-paragraph-3']")
            )
        ).text

        pattern_total_num = 'Showing.*of\s(\d*).*'
        pattern_num_per_page = 'Showing\s\d*\s-\s(\d*)\sof\s\d*.*'
        result['total_num'] = re.search(
            pattern_total_num, totalProductsOutput).group(1)
        result['num_per_page'] = re.search(
            pattern_num_per_page, totalProductsOutput).group(1)
    except:
        return False
    finally:
        return result

# get all Laptops New sku items


def get_sku_items(driver, link, pages_num):
    skip_items = 0
    driver.get(link+str(skip_items))

    close_dialog(driver)
    # get items for each page
    for i in range(pages_num):
        try:
            sku_items = WebDriverWait(driver, 10).until(
                EC.presence_of_all_elements_located(
                    (By.XPATH, "//div[@class='m-channel-placement-item f-wide f-full-bleed-image']/a"))
            )
            cur_item_list = get_cur_page_items(sku_items)
        except:
            return False

        # if has next page, click next and wait until page refresh
        has_next_page = i < pages_num - 1
        if has_next_page:
            seed()
            time.sleep(randint(10, 15))
            navigate_next_page(driver, link, i)
            wait_until_page_refresh(driver, cur_item_list)


def close_dialog(driver):
    try:
        dialog_close_btn = WebDriverWait(driver, 10).until(
            EC.element_to_be_clickable(
                (By.XPATH,
                 "//div[@class='sfw-dialog']/div[@class='c-glyph glyph-cancel']")
            )
        )
        dialog_close_btn.click()
    except:
        return


def get_cur_page_items(sku_items):
    cur_item_list = []

    for item_element in sku_items:
        item = {}
        data = json.loads(item_element.get_attribute("data-m"))
        prd_id = data['pid']
        prd_name = data['tags']['prdName']

        item['link'] = 'https://www.microsoft.com/en-us/d/' + \
            prd_name.replace(" ", "-").replace('"', "").lower()+'/'+prd_id
        item['sku'] = prd_id
        item['currentPrice'] = get_sku_item_price(item_element)
        item['name'] = prd_name

        # print out item, caught by process listen data.
        print(json.dumps(item))
        cur_item_list.append(prd_id)

    return cur_item_list


def get_sku_item_price(driver):
    price = None
    try:
        price_element = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located(
                (By.CSS_SELECTOR, "span[itemprop='price']")
            )
        )
        price = price_element.get_attribute(
            "content").strip().lstrip("$").replace(',', '')
    except:
        return False
    finally:
        return price


def navigate_next_page(driver, link, index):
    editURL = link + str((index+1)*20)
    try:
        driver.get(editURL)
    except:
        return False


def wait_until_page_refresh(driver, cur_item_list):
    try:
        WebDriverWait(driver, 20).until(
            (lambda x: sku_attribute_changed(driver, cur_item_list))
        )
        return True
    except:
        return False


def sku_attribute_changed(driver, skus_before_changed):
    try:
        sku_items = WebDriverWait(driver, 20).until(
            EC.presence_of_all_elements_located(
                (By.XPATH,
                 "//div[@class='m-channel-placement-item f-wide f-full-bleed-image']/a")
            )
        )
        index = 0
        for item_element in sku_items:
            data = json.loads(item_element.get_attribute("data-m"))
            item_sku = data['pid']

            # check if list elements being refreshed
            if(item_sku == skus_before_changed[index] or item_sku == None):
                # print(f'sku_item not refreshed')
                return False
            index += 1
        return sku_items
    except:
        return False

File: server/bin/helper/MS.js
import { AlertApi } from '../../lib/query/utilities.js';
import Stores from './Stores.js';

/*
declare class Microsoft{
    initURL(skipItemsNum):url
    async getItemSpec(page:puppeteer<page>, url:URL): ItemSpec
    async #openSpecWrapper(page:puppeteer<page>):void
    async #parseItemSpec(page:puppeteer<page>): ItemSpec
    async #parsePageNumFooter(page:puppeteer<page>): PageNumFooter
    async closeDialog(page:puppeteer<page>):void
    async getPagesNum(page:puppeteer<page>, url:URL): PageNumFooter>
    async #parseItemsList(page:puppeteer<page>): Array<Item>
    async getPageItems(page:puppeteer<page>, url:URL): Array<Item>
    
}
*/
/* 
interface Item{
    link: url,
    sku: string,
    currentPrice: string,
    name: string
}
*/
/* 
interface ItemSpec{
    [key:string]:value:string
    ...
}
*/

/* 
interface PageNumFooter{
    numPerPage: number,
    totalNum: number
}

*/

export default class Microsoft extends Stores {
    constructor() {
        super();
        this.url = 'https://www.microsoft.com/en-us/store/b/shop-all-pcs?categories=2+in+1||Laptops||Desktops||PC+Gaming&s=store&skipitems=';
    }

    async getAndSaveMicrosoftLaptopsPrice() {
        let erpApi = new AlertApi();
        let databaseModel = erpApi.getMicrosoftAlertModel();
        let storeUrl = this.initURL(0); //this. url + skipItemsNum

        let browser = await this.initBrowser();
        let page = await this.initPage(browser);
        await page.setDefaultNavigationTimeout(30000);

        let { pagesNum, numPerPage } = await this.getPagesNum(page, storeUrl);
        console.log('total pages num: ', pagesNum);
        try {
            for (let i = 0; i < pagesNum; i++) {
                let pageUrl = this.initURL(i * numPerPage);
                let promisesRes = await this.getPageItems(page, pageUrl); //Array<{PromiseResolveType}>
                let details = promisesRes.filter(prom => prom.status === 'fulfilled' && prom.value.price !== undefined)
                await Promise.all(
                    details.map((detail, index) => {
                        const { pid, cN, price } = detail.value;
                        let itemDetail = {
                            sku: pid,
                            name: cN,
                            link: 'https://www.microsoft.com/en-us/d/' + cN.replace(/\s/g, "-").replace(/"/g, "").toLowerCase() + '/' + pid,
                            currentPrice: price
                        }
                        erpApi.saveStoreItemToDatabase(itemDetail, databaseModel).then(result => {
                            this.printMsg(new Map(Object.entries({
                                store: Microsoft.name,
                                page: i,
                                index: index,
                                sku: itemDetail.sku,
                                currentPrice: itemDetail.currentPrice,
                                result: result
                            })))
                        });
                    }))
                    .finally(console.log(`[${Microsoft.name}]===Page ${i} finished.===`))
            }

            await page.close();
            await browser.close();
        } catch (e) {
            await page.close();
            await browser.close()
            console.error(`\nERROR:[${Microsoft.name}] Ended with exception.\n`, e.message)
            throw new Error(e.message);
        }

    }

    initURL(skipItemsNum) {
        return this.url + skipItemsNum
    }

    async #parsePageNumFooter(page) {
        let res = {
            numPerPage: undefined,
            totalNum: undefined
        };
        const FOOTER_XPATH_EXPR = '//p[@class="c-paragraph-3"]'
        const NUM_PAGE_REGEX_EXPR = /.*Showing\s\d*\s-\s(\d*)\sof\s\d*.*/
        const TOTAL_NUM_REGEX_EXPR = /.*Showing.*of\s(\d*).*/

        let footer = (await this.evaluateElementsText(page, FOOTER_XPATH_EXPR))[0]
        res.numPerPage = Number(this.getRegexValue(footer, NUM_PAGE_REGEX_EXPR))
        res.totalNum = Number(this.getRegexValue(footer, TOTAL_NUM_REGEX_EXPR))

        console.log(`[${this.constructor.name}][Parse Page Num Footer] numPerPage:${res.numPerPage}, totalNum:${res.totalNum}`)
        return res
    }

    async #closeDialogIfAny(page) {
        try {
            await page.waitForXPath('//div[@class="sfw-dialog"]/div[@class="c-glyph glyph-cancel"]')
            let dialogCloseBtn = (await page.$x('//div[@class="sfw-dialog"]/div[@class="c-glyph glyph-cancel"]'))[0]
            await dialogCloseBtn.click()
        } catch (e) {
            console.error(`${this.constructor.name}: no dialog shows up, skipping.`)
        }
    }

    /* 
    @param: page:Puppeteer<page>
    @param: url: string
    @return: PageNumFooter
    */
    async getPagesNum(page, url) {
        await page.goto(url);
        await this.#closeDialogIfAny(page)    //may or may not close the dialog, it depends if the dialog shows up.
        let res = await this.#parsePageNumFooter(page)

        let pageNumFooter = {
            pagesNum: Math.ceil(res.totalNum / res.numPerPage),
            numPerPage: res.numPerPage
        }

        return pageNumFooter
    }

    /* 
        @param: page:Puppeteer<page>
        @param: url: string
        @return: Array<Item>
    */
    async #parseItemsList(page) {
        const ITEM_ELEMENTS_EXPR = '//div[@class="m-channel-placement-item f-wide f-full-bleed-image"]'
        const PRICE_SPAN_EXPR = 'span[itemprop="price"]'
        const IS_INSTOCK_EXPR = 'strong[class="c-badge f-small f-lowlight x-hidden-focus"]'
        const ITEM_ATTRIBUTE_ID = "data-m"
        const PRICE_ATTRIBUTE_ID = "content"

        // let itemAttrLists = await this.evaluateItemAttribute(page, ITEMS_LIST_EXPR, ITEM_ATTRIBUTE_ID)
        // let priceAttrLists = await this.evaluatePriceAttribute(page, PRICE_LIST_EXPR, PRICE_ATTRIBUTE_ID)

        await page.waitForXPath(ITEM_ELEMENTS_EXPR);
        let itemElements = await page.$x(ITEM_ELEMENTS_EXPR);
        // let itemElements = await page.$$('div.m-channel-placement-item f-wide f-full-bleed-image');

        let results = await Promise.allSettled(itemElements.map(async (ele) => {
            let price = undefined, isInStock = true;
            let attrRes = await ele.$eval('a', (ele, ITEM_ATTRIBUTE_ID) => ele.getAttribute(ITEM_ATTRIBUTE_ID), ITEM_ATTRIBUTE_ID)
            let data = JSON.parse(attrRes)//JSON text data-m attribute to JSON object
            try {
                let priceRes = await ele.$eval(PRICE_SPAN_EXPR, (span, PRICE_ATTRIBUTE_ID) => span.getAttribute(PRICE_ATTRIBUTE_ID), PRICE_ATTRIBUTE_ID)
                price = Number(priceRes.replace(/[$|,]/g, ""));
            } catch {
                return { ...data, price, isInStock: false } //if no price found, not in stock
            }
            try {
                isInStockRes = await ele.$eval(IS_INSTOCK_EXPR, el => el.textContent)
                isInStock = isInStockRes === "OUT OF STOCK" ? false : true;
            } catch {
                //if no OUT OF STOCK tag do nothing...
            }
            return { ...data, price, isInStock };   //found price and currently instock
        }))

        return results;
    }
    /* 
        @param: page:Puppeteer<page>
        @param: url: string
        @return: Array<Item>
    */
    async getPageItems(page, url) {
        await page.goto(url)
        // await page.waitForTimeout(10000);

        let items = await this.#parseItemsList(page)
        return items
    }


    async getItemSpec(page, url) {
        return
    }
}
